agent/agent/src/main/java/com/microsoft/applicationinsights/agent/Main.java

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.applicationinsights.agent;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.CodeSource;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.jar.Attributes;
+import java.util.jar.JarEntry;
+import java.util.jar.JarInputStream;
+import java.util.jar.JarOutputStream;
+import java.util.jar.Manifest;
+import java.util.regex.Pattern;
+
+@SuppressWarnings("SystemOut")
+public class Main {
+
+  private static final Pattern SIGNATURE_FILE = Pattern.compile("META-INF/.*\\.(RSA|SF)");
+
+  public static void main(String[] args) throws Exception {
+    if (args.length == 0) {
+      System.err.println("No command specified");
+      return;
+    }
+    String command = args[0];
+    if (command.equals("unsign")) {
+      if (args.length > 1) {
+        System.err.println("The unsign command does not expect any arguments");
+        return;
+      }
+      unsign();
+      return;
+    }
+    System.err.println("Unknown command: " + command);
+  }
+
+  private static void unsign() throws IOException, URISyntaxException {
+    Path signedPath = getAgentJar();
+    Path unsignedPath = getUnsignedPath(signedPath);
+
+    ScheduledExecutorService executor;
+    try (JarInputStream in = new JarInputStream(Files.newInputStream(signedPath))) {
+
+      Manifest signedManifest = in.getManifest();
+
+      if (!isSigned(signedManifest)) {
+        System.err.println();
+        System.err.println(getRelativePath(signedPath) + " is already unsigned");
+        in.close();
+        System.exit(0);
+      }
+
+      executor = Executors.newSingleThreadScheduledExecutor();
+
+      System.out.println();
+      System.out.print("Unsigning...");
+      System.out.flush();
+      executor.scheduleWithFixedDelay(
+          () -> {
+            System.out.print(".");
+            System.out.flush();
+          },
+          500,
+          500,
+          MILLISECONDS);
+
+      unsign(in, signedManifest, unsignedPath);
+    }
+
+    shutdown(executor);
+
+    System.err.println(); // for the "Unsigned......." line
+
+    System.out.println();
+    System.out.println("Unsigned agent jar: " + getRelativePath(unsignedPath));
+  }
+
+  private static void unsign(JarInputStream in, Manifest signedManifest, Path unsignedPath)
```

Comment:
The responsability of the Main class seems to execute one action type according to a command value (strategy pattern). That would be its only responsibility for me and it would not be necessary that the Main class knows the implementation details of the jar unsigning.

I would have put the code related to the jar unsigning in a specific class.  

If one day we add a feature in the main class, we could immediatly get the idea to create another class and it would help to avoid having a big Main class in the end doing many different things.

Reply:
done 👍